// framework7-react/components/breadcrumbs-item.js
function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

import React, { forwardRef, useRef, useImperativeHandle } from 'react';
import { classNames, getExtraAttrs, emit } from '../shared/utils.js';

const BreadcrumbsItem = /*#__PURE__*/forwardRef((props, ref) => {
  const {
    className,
    id,
    style,
    active,
    children
  } = props;
  const extraAttrs = getExtraAttrs(props);
  const elRef = useRef(null);
  useImperativeHandle(ref, () => ({
    el: elRef.current
  }));

  const onClick = e => {
    emit(props, 'click', e);
  };

  return /*#__PURE__*/React.createElement("div", _extends({
    className: classNames('breadcrumbs-item', className, active && 'breadcrumbs-item-active'),
    ref: elRef,
    id: id,
    style: style,
    onClick: onClick
  }, extraAttrs), children);
});
BreadcrumbsItem.displayName = 'f7-breadcrumbs-item';
export default BreadcrumbsItem;
