// framework7-react/components/checkbox.js
function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

import React, { forwardRef, useRef, useImperativeHandle, useEffect } from 'react';
import { classNames, getExtraAttrs, emit } from '../shared/utils.js';
import { colorClasses } from '../shared/mixins.js';

const Checkbox = /*#__PURE__*/forwardRef((props, ref) => {
  const {
    className,
    id,
    style,
    children,
    name,
    value,
    disabled,
    readonly,
    checked,
    defaultChecked,
    indeterminate
  } = props;
  const extraAttrs = getExtraAttrs(props);
  const elRef = useRef(null);
  const inputElRef = useRef(null);
  useImperativeHandle(ref, () => ({
    el: elRef.current,
    inputEl: inputElRef.current
  }));

  const onChange = event => {
    emit(props, 'change', event);
  };

  useEffect(() => {
    if (inputElRef.current) {
      inputElRef.current.indeterminate = !!indeterminate;
    }
  }, [indeterminate]);
  const inputEl = /*#__PURE__*/React.createElement("input", {
    ref: inputElRef,
    type: "checkbox",
    name: name,
    value: value,
    disabled: disabled,
    readOnly: readonly,
    checked: checked,
    defaultChecked: defaultChecked,
    onChange: onChange
  });
  const iconEl = /*#__PURE__*/React.createElement("i", {
    className: "icon-checkbox"
  });
  const classes = classNames(className, {
    checkbox: true,
    disabled
  }, colorClasses(props));
  return /*#__PURE__*/React.createElement("label", _extends({
    id: id,
    style: style,
    className: classes,
    ref: elRef
  }, extraAttrs), inputEl, iconEl, children);
});
Checkbox.displayName = 'f7-checkbox';
export default Checkbox;
