// framework7-react/components/messagebar-attachment.js
function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

import React, { forwardRef, useRef, useImperativeHandle } from 'react';
import { classNames, getExtraAttrs, emit } from '../shared/utils.js';
import { colorClasses } from '../shared/mixins.js';

const MessagebarAttachment = /*#__PURE__*/forwardRef((props, ref) => {
  const {
    className,
    id,
    style,
    children,
    image,
    deletable = true
  } = props;
  const extraAttrs = getExtraAttrs(props);
  const elRef = useRef(null);

  const onClick = event => {
    emit(props, 'attachmentClick', event);
  };

  const onDeleteClick = event => {
    emit(props, 'attachmentDelete', event);
  };

  useImperativeHandle(ref, () => ({
    el: elRef.current
  }));
  const classes = classNames(className, 'messagebar-attachment', colorClasses(props));
  return /*#__PURE__*/React.createElement("div", _extends({
    id: id,
    style: style,
    className: classes,
    ref: elRef
  }, extraAttrs, {
    onClick: onClick
  }), image && /*#__PURE__*/React.createElement("img", {
    src: image
  }), deletable && /*#__PURE__*/React.createElement("span", {
    className: "messagebar-attachment-delete",
    onClick: onDeleteClick
  }), children);
});
MessagebarAttachment.displayName = 'f7-messagebar-attachment';
export default MessagebarAttachment;
