// framework7-react/components/messagebar-sheet-image.js
function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

import React, { forwardRef, useRef, useImperativeHandle } from 'react';
import { classNames, getExtraAttrs, emit } from '../shared/utils.js';
import { colorClasses } from '../shared/mixins.js';

const MessagebarSheetImage = /*#__PURE__*/forwardRef((props, ref) => {
  const {
    className,
    id,
    style,
    children,
    image,
    checked
  } = props;
  const extraAttrs = getExtraAttrs(props);
  const elRef = useRef(null);

  const onChange = event => {
    if (event.target.checked) emit(props, 'checked', event);else emit(props, 'unchecked', event);
    emit(props, 'change', event);
  };

  useImperativeHandle(ref, () => ({
    el: elRef.current
  }));
  const classes = classNames(className, 'messagebar-sheet-image', 'checkbox', colorClasses(props));
  const styles = {
    backgroundImage: image && `url(${image})`,
    ...(style || {})
  };
  return /*#__PURE__*/React.createElement("label", _extends({
    id: id,
    className: classes,
    style: styles,
    ref: elRef
  }, extraAttrs), /*#__PURE__*/React.createElement("input", {
    type: "checkbox",
    checked: checked,
    onChange: onChange
  }), /*#__PURE__*/React.createElement("i", {
    className: "icon icon-checkbox"
  }), children);
});
MessagebarSheetImage.displayName = 'f7-messagebar-sheet-image';
export default MessagebarSheetImage;
