// framework7-react/components/nav-left.js
function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

import React, { forwardRef, useRef, useImperativeHandle } from 'react';
import { classNames, getExtraAttrs, emit } from '../shared/utils.js';
import { colorClasses } from '../shared/mixins.js';
import { useTheme } from '../shared/use-theme.js';
import Link from './link.js';

const NavLeft = /*#__PURE__*/forwardRef((props, ref) => {
  const {
    className,
    id,
    style,
    children,
    backLink,
    backLinkUrl,
    backLinkForce,
    backLinkShowText,
    sliding
  } = props;
  const extraAttrs = getExtraAttrs(props);
  const elRef = useRef(null);

  const onBackClick = event => {
    emit(props, 'backClick clickBack', event);
  };

  useImperativeHandle(ref, () => ({
    el: elRef.current
  }));
  const theme = useTheme();
  let linkEl;
  let needBackLinkText = backLinkShowText;
  if (typeof needBackLinkText === 'undefined') needBackLinkText = !theme.md;

  if (backLink) {
    const text = backLink !== true && needBackLinkText ? backLink : undefined;
    linkEl = /*#__PURE__*/React.createElement(Link, {
      href: backLinkUrl || '#',
      back: true,
      icon: "icon-back",
      force: backLinkForce || undefined,
      className: !text ? 'icon-only' : undefined,
      text: text,
      onClick: onBackClick
    });
  }

  const classes = classNames(className, 'left', {
    sliding
  }, colorClasses(props));
  return /*#__PURE__*/React.createElement("div", _extends({
    id: id,
    style: style,
    className: classes,
    ref: elRef
  }, extraAttrs), linkEl, children);
});
NavLeft.displayName = 'f7-nav-left';
export default NavLeft;
