// framework7-react/shared/router-open-in.js
import React from 'react';
import Popup from '../components/popup.js';
import View from '../components/view.js';
import LoginScreen from '../components/login-screen.js';
import Sheet from '../components/sheet.js';
import Popover from '../components/popover.js';
import Panel from '../components/panel.js';
export const routerOpenIn = (router, url, options) => {
  const navigateOptions = {
    url,
    route: {
      path: url,
      options: { ...options,
        openIn: undefined
      }
    }
  };
  const params = { ...options
  };

  if (options.openIn === 'popup') {
    params.component = function DynamicPopup() {
      return /*#__PURE__*/React.createElement(Popup, {
        className: "popup-router-open-in",
        "data-url": url
      }, /*#__PURE__*/React.createElement(View, {
        linksView: router.view.selector,
        url: url,
        ignoreOpenIn: true
      }));
    };

    navigateOptions.route.popup = params;
  }

  if (options.openIn === 'loginScreen') {
    params.component = function DynamicPopover() {
      return /*#__PURE__*/React.createElement(LoginScreen, {
        className: "login-screen-router-open-in",
        "data-url": url
      }, /*#__PURE__*/React.createElement(View, {
        linksView: router.view.selector,
        url: url,
        ignoreOpenIn: true
      }));
    };

    navigateOptions.route.loginScreen = params;
  }

  if (options.openIn === 'sheet') {
    params.component = function DynamicSheet() {
      return /*#__PURE__*/React.createElement(Sheet, {
        className: "sheet-modal-router-open-in",
        "data-url": url
      }, /*#__PURE__*/React.createElement(View, {
        linksView: router.view.selector,
        url: url,
        ignoreOpenIn: true
      }));
    };

    navigateOptions.route.sheet = params;
  }

  if (options.openIn === 'popover') {
    params.targetEl = options.clickedEl || options.targetEl;

    params.component = function DynamicPopover() {
      return /*#__PURE__*/React.createElement(Popover, {
        className: "popover-router-open-in",
        targetEl: options.clickedEl || options.targetEl,
        "data-url": url
      }, /*#__PURE__*/React.createElement(View, {
        linksView: router.view.selector,
        url: url,
        ignoreOpenIn: true
      }));
    };

    navigateOptions.route.popover = params;
  }

  if (options.openIn.indexOf('panel') >= 0) {
    const parts = options.openIn.split(':');
    const side = parts[1] || 'left';
    const effect = parts[2] || 'cover';

    params.component = function DynamicPanel() {
      return /*#__PURE__*/React.createElement(Panel, {
        side: side,
        effect: effect,
        className: "panel-router-open-in",
        "data-url": url
      }, /*#__PURE__*/React.createElement(View, {
        linksView: router.view.selector,
        url: url,
        ignoreOpenIn: true
      }));
    };

    navigateOptions.route.panel = params;
  }

  return router.navigate(navigateOptions);
};
