// @deck.gl/core/typed/lib/layer-state.js
import ComponentState from '../lifecycle/component-state';
export default class LayerState extends ComponentState {
    constructor({ attributeManager, layer }) {
        super(layer);
        this.attributeManager = attributeManager;
        this.needsRedraw = true;
        this.needsUpdate = true;
        this.subLayers = null;
        this.usesPickingColorCache = false;
    }
    get layer() {
        return this.component;
    }
    set layer(layer) {
        this.component = layer;
    }
    /* Override base Component methods with Layer-specific handling */
    _fetch(propName, url) {
        const fetch = this.component.props.fetch;
        if (fetch) {
            return fetch(url, { propName, layer: this.layer });
        }
        return super._fetch(propName, url);
    }
    _onResolve(propName, value) {
        const onDataLoad = this.component.props.onDataLoad;
        if (propName === 'data' && onDataLoad) {
            onDataLoad(value, { propName, layer: this.layer });
        }
    }
    _onError(propName, error) {
        this.layer.raiseError(error, `loading ${propName} of ${this.layer}`);
    }
}
